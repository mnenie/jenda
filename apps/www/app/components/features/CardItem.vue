<script lang="ts">
export interface CardItem {
  title: string
  description: string
  src: string
  imgClass?: HTMLAttributes['class']
}
</script>

<script setup lang="ts">
import type { HTMLAttributes } from 'vue'

defineProps<{
  card: CardItem
}>()
</script>

<template>
  <UnaCard
    :title="card.title"
    :description="card.description"
    card="soft-gray"
    class="shadow-none w-full min-h-260px h-full relative overflow-hidden grow flex flex-col"
    :una="{
      cardTitle: 'text-2xl font-semibold',
      cardDescription: 'text-md text-neutral-500',
      cardContent: 'flex flex-col justify-between h-full',
    }"
  >
    <NuxtImg :src="card.src" class="mt-auto max-w-unset" :class="card.imgClass" rounded="xl" />
  </UnaCard>
</template>
