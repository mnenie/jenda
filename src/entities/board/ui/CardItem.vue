<script setup lang="ts">
import { computed } from 'vue';
import { UserAvatar } from '@/entities/user';
import { UiBadge, UiCard } from '@/shared/ui';
import { MessagesSquare } from 'lucide-vue-next';
import type { Card } from '../model';

const props = defineProps<{
  card: Card;
}>();

const colors = ['low', 'medium', 'high', 'none'] as const

const priority = computed(() => {
  const index = colors.indexOf(props.card.priority)
  return index !== -1 && index !== 3 ? `var(--${colors[index]}-color)` : 'white'
});
</script>

<template>
  <UiCard :style="{ backgroundColor: `rgba(${priority}, 0.2)` }" :class="$style.card">
    <div :class="$style.top">
      <div :class="$style.name">
        <span>{{ card.title }}</span>
      </div>
      <div v-if="card.tags?.length !== 0" :class="$style.tags">
        <UiBadge v-for="tag in card.tags" :key="tag._id" variant="secondary" style="width: fit-content">
          {{ tag.name }}
        </UiBadge>
      </div>
    </div>
    <div :class="$style.bottom">
      <div :class="$style.left_container">
        <span> #{{ card._id }}</span>
        <div v-if="card.chat" :class="$style.messages">
          <MessagesSquare :size="14" />
          <span>{{ card.chatCount }}</span>
        </div>
      </div>
      <div :class="$style.right_container">
        <div :class="$style.user_container">
          <UserAvatar v-for="user in card.users" :key="user._id" :class="$style.user">
            <img :src="user.photoUrl" :alt="user.email" style="width: 100%" />
          </UserAvatar>
        </div>
      </div>
    </div>
    <div
      v-if="card.priority !== 'none'"
      :style="{ backgroundColor: `rgb(${priority})` }"
      :class="$style.indicator"
    />
  </UiCard>
</template>

<style module lang="scss">
@import '@/app/styles/mixins';

.card {
  cursor: pointer;
  @include transition;
  @include on-hover {
    background-color: var(--zinc-50);
  }

  .top {
    display: flex;
    flex-direction: column;
    gap: 8px;

    .tags {
      display: flex;
      align-items: center;
      gap: 4px;
    }
  }

  .name {
    display: flex;
    align-items: center;
    justify-content: space-between;

    & > span {
      max-width: 200px;
    }
  }

  .user_container {
    display: flex;
    align-items: center;
    gap: 4px;
    height: 21px;

    .user {
      width: 21px;
      height: 21px;
    }
  }

  .indicator {
    width: 20px;
    height: 4px;
    position: absolute;
    top: 0;
    right: 10px;
    border-radius: 0px 0px 4px 4px;
  }

  .bottom {
    margin-top: 12px;
    display: flex;
    align-items: center;
    justify-content: space-between;

    .left_container {
      display: flex;
      align-items: center;
      gap: 8px;

      & > span {
        color: var(--zinc-400);
      }

      .messages {
        display: flex;
        align-items: center;
        gap: 1px;
        color: var(--zinc-400);

        & > span {
          font-size: 11px;
        }
      }
    }

    .right_container {
      display: flex;
      align-items: center;
      gap: 8px;
    }
  }
}

:global(html.dark) {
  .card {
    background-color: rgba(var(--zinc-rgb-600), 0.4);
    border-color: var(--zinc-600);
    color: var(--zinc-100);
    .bottom {
      .left_container {
        & > span {
          color: var(--zinc-300);
        }
        .messages {
          color: var(--zinc-300);
        }
      }
    }
  }
}
</style>
