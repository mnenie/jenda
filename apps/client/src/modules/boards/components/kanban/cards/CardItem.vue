<script setup lang="ts" generic="U extends Extract<T['priority'], string>, T extends BoardCard">
import { computed } from 'vue'
import VisibleLabels from '../../additions/VisibleLabels.vue'
import CardBottomPanel from './CardBottomPanel.vue'
import type { BoardCard, Priority } from '@/modules/boards/types'
import { UiButton } from '@/shared/ui'
import { cn } from '@/shared/libs/shadcn/utils'

const props = defineProps<{
  card: T
  subtasks?: Pick<T, 'subtasks'>[]
}>()

const priorities = {
  none: '',
  low: '!task-card-green',
  medium: '!task-card-yellow',
  high: '!task-card-red',
} as const satisfies Record<U | Priority, string>

const priorityColor = computed(() => priorities[props.card.priority!])
</script>

<template>
  <div
    draggable="true"
    :class="cn(
      'relative z-99 w-full cursor-grab active:cursor-grabbing p-2.5 rounded-md border border-neutral-200 bg-white flex flex-col items-start dark:(border-neutral-700 bg-#2e2e2e)',
      priorityColor && 'border-none',
    )"
  >
    <div
      v-if="priorityColor"
      :class="cn(
        'absolute z--1 inset-0 rounded-md pointer-events-none w-full h-full border after:absolute after:inset-0',
        priorityColor,
      )"
    />
    <div v-if="card.labels?.length" class="flex items-center justify-between gap-2 w-full mb-1.5">
      <VisibleLabels :labels="card.labels" :max="3" class="gap-2" label-class="!max-w-40" />
    </div>
    <div class="flex flex-col w-full gap-1 mb-4">
      <p class="text-default font-medium mb-1 truncate">
        {{ card.title }}
      </p>
      <div class="flex items-stretch gap-1.5 w-full h-fit ml-0.5">
        <div class="min-w-2px bg-neutral-200 rounded-xl dark:bg-neutral-600" />
        <div class="flex flex-col gap-1 w-full">
          <UiButton
            v-for="subtask in card.subtasks"
            :key="subtask._id"
            variant="ghost"
            class="!h-7 w-full justify-start !text-neutral-500"
          >
            <span class="truncate dark:text-neutral-300">{{ subtask.title }}</span>
          </UiButton>
        </div>
      </div>
    </div>
    <CardBottomPanel :card />
  </div>
</template>
