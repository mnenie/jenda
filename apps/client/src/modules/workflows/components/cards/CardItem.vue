<script setup lang="ts" generic="T extends WorkflowCard">
import { computed, inject } from 'vue'
import type { WorkflowCard } from '../../types'
import { UiBadge } from '@/shared/ui'
import { DayjsInjectionKey } from '@/plugins/dayjs'

defineProps<{
  card: T
}>()

const dayjs = inject(DayjsInjectionKey)!

const timesAgo = computed(() =>
  (workflow: T) => {
    return dayjs().to(dayjs(workflow.updatedAt))
  })
</script>

<template>
  <div
    class="relative p-3 cursor-pointer w-full flex flex-col justify-between h-full border border-solid border-neutral-200 rounded-8px dark:border-neutral-700/60"
    @click="$router.push({ name: 'workflows-id', params: { id: card._id! } })"
  >
    <div class="absolute inset-0 bg-dotted-pattern opacity-30 pointer-events-none" />

    <div class="w-full flex items-start justify-between">
      <div class="flex flex-col gap-0.5 w-full">
        <p class="text-default hyphens-auto">
          {{ card.name }}
        </p>
        <p class="text-small text-neutral-500 hyphens-auto overflow-hidden text-ellipsis line-clamp-1">
          {{ card.description }}
        </p>
      </div>
      <UiBadge variant="soft" class="px-1 py-px mt-px" :class="card.state === 'production' ? 'badge-soft-green' : 'badge-soft'">
        {{ $t(`workflows.status.${card.state}`) }}
      </UiBadge>
    </div>
    <span class="text-small mt-3 text-neutral-500 overflow-hidden text-ellipsis line-clamp-1 dark:text-neutral-400">
      {{ timesAgo(card) }}
    </span>
  </div>
</template>

<style scoped>
.bg-dotted-pattern {
  background-image: radial-gradient(circle, #ccc 1px, transparent 1px);
  background-size: 12px 12px;
  background-color: inherit;
}

:global(html.dark) {
  .bg-dotted-pattern {
    background-image: radial-gradient(circle, #424242 1px, transparent 1px);
  }
}
</style>
