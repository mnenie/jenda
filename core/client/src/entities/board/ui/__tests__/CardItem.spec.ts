import { mount } from '@vue/test-utils'
import { describe, expect, it } from 'vitest'
import CardItem from '../CardItem.vue'

describe('tests for CardItem.vue', () => {
  const wrapper = mount(CardItem, {
    props: {
      card: {
        _id: '0',
        title: 'Test Card',
        priority: 'low',
        users: [],
        chat: true,
        chatCount: 2,
        tags: [{ _id: '0', name: 'test' }],
      },
    },
  })

  it('should render correctly', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should has correct number of tags', () => {
    const tags = wrapper.findAll('.tags > .badge')
    expect(tags.length).toBe(1)
    expect(tags[0].text()).toBe('test')
  })

  it('should has correct chat counts', () => {
    const chat = wrapper.find('.messages > span')
    expect(chat.text()).toBe('2')
  })
})
