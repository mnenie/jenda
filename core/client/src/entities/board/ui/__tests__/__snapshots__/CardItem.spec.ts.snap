// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`tests for CardItem.vue > should render correctly 1`] = `
"<div class="text-sm card card">
  <div class="top">
    <div class="name"><span>Test Card</span></div>
    <div class="tags">
      <div class="text-xs badge secondary">test</div>
    </div>
  </div>
  <div class="bottom">
    <div class="left_container"><span> #0</span>
      <div class="messages">
        <div i-hugeicons-message-multiple-02="" text-sm=""></div><span>2</span>
      </div>
    </div>
    <div class="right_container">
      <div class="user_container"></div>
    </div>
  </div>
  <div class="indicator"></div>
  <div class="user_container"></div>
</div>"
`;
